criu/proc_parse.c

```diff
@@ -1527,6 +1527,56 @@ int parse_timens_offsets(struct timespec *boff, struct timespec *moff)
 	return exit_code;
 }
 
+static int get_mountinfo_sdev_from_mntid(int mnt_id, unsigned int *sdev)
+{
+	int exit_code = -1;
+	FILE *f;
+
+	f = fopen_proc(PROC_SELF, "mountinfo");
```

Comment:
can we read criu's mountiunfo once?

Reply:
Thanks for this question!

Disclamer: It looks like I had the right idea but implemented it wrong.

We have one code path where __open_mountpoint() can be called with mnt_fd != -1 in open_mountpoint(). In all other code paths we just open mnt_fd ourselfs as pm->ns_mountpoint at mntns_root, so fdinfo mount id would be from same mntns as pm, thus always
present in mountinfo.

But in open_mountpoint we can open fd via: 
1) get_clean_fd which would give fd with mnt_id of newly created clean bindmount, which we don't have in mountinfo
(patch logic with reading mountinfo should not be helping in this case as before __open_mountpoint we do restore_mnt_ns =( )
2) ns_open_mountpoint which would give fd with mnt_id from newly cloned helper-mntns, which we also don't have in mountinfo

So if we dump overmounted btrfs volume mount (which we don't do in tests, that's why I didn't initially found it) __open_mountp
oint would still fail to validate mnt_fd.

I will think some more on how to fix it, if have any ideas - please share them.